File: Python/khan.py
from collections import deque

def topological_sort_kahn(graph):
    indegree = {u: 0 for u in graph}

    for u in graph:
        for v in graph[u]:
            indegree[v] += 1

    queue = deque([u for u in graph if indegree[u] == 0])
    topo_order = []

    while queue:
        u = queue.popleft()
        topo_order.append(u)

        for v in graph[u]:
            indegree[v] -= 1
            if indegree[v] == 0:
                queue.append(v)

    if len(topo_order) != len(graph):
        raise ValueError("Graph has a cycle")

    return topo_order


# Example usage
graph = {
    'A': ['B', 'C'],
    'B': ['D'],
    'C': ['D'],
    'D': []
}
try:
    order = topological_sort_kahn(graph)
    print("Topological Sort Order:", order)
except ValueError as e:
    print(e)

File: TypeScript/khan.ts
/**
 * Implementation of Kahn's Algorithm for Topological Sort
 */
function topologicalSortKahn(graph: Record<string, string[]>): string[] {
    const indegree: Record<string, number> = {};
    const nodes = Object.keys(graph);

    // Initialize indegrees to 0
    for (const u of nodes) {
        indegree[u] = 0;
    }

    // Calculate indegree for each node
    for (const u of nodes) {
        for (const v of graph[u]) {
            indegree[v] = (indegree[v] || 0) + 1;
        }
    }

    // Initialize queue with nodes having 0 indegree
    const queue: string[] = nodes.filter(u => indegree[u] === 0);
    const topoOrder: string[] = [];

    while (queue.length > 0) {
        // shift() mimics deque.popleft()
        const u = queue.shift()!;
        topoOrder.push(u);

        for (const v of graph[u]) {
            indegree[v]--;
            if (indegree[v] === 0) {
                queue.push(v);
            }
        }
    }

    if (topoOrder.length !== nodes.length) {
        throw new Error("Graph has a cycle");
    }

    return topoOrder;
}

// Example usage
const graph2: Record<string, string[]> = {
    'A': ['B', 'C'],
    'B': ['D'],
    'C': ['D'],
    'D': []
};

try {
    const order = topologicalSortKahn(graph2);
    console.log("Topological Sort Order:", order);
} catch (e) {
    if (e instanceof Error) {
        console.error(e.message);
    }
}
